[Blog Post] Tables are Hard, Part 2: Building a Simple Data Table in React (#234)
pixie-io/blog.px.dev: content/blog/tables-are-hard-2/index.md

---
path: '/tables-are-hard-2'
title: 'Tables are Hard, Part 2: Building a Simple Data Table in React'
date: 2021-12-14T06:00:00.000+00:00
featured_image: tables-are-hard-2-hero.png
categories: ['Pixie Team Blogs']
authors: ['Nick Lanam']
emails: ['nlanam@pixielabs.ai']
---

Let's build a "simple" data table in a web app, starting with the basics.

In the [previous post](https://blog.px.dev/tables-are-hard-1/), we summarized the features expected in a modern data table.

By the end of this tutorial, we'll build a table with filtering, sorting, and basic column controls.
In the next post, we'll add new requirements and walk through handling them.

This tutorial is aimed at relative beginners to web development, and will keep the set of tools as basic as possible to remain focused. Skip to the end for suggestions for a more serious project.

> [Final Code](https://github.com/pixie-io/pixie-demos/react-table-demo) | [Live Demo](https://pixie-io.github.io/pixie-demos/react-table-demo)

Reviewer: ```suggestion
[Final Code](https://github.com/pixie-io/pixie-demos/react-table) | [Live Demo](https://pixie-io.github.io/pixie-demos/react-table)
```

Author: Oops, nice catch! Done.

+  opacity: 0.8;\n\
+}\n\
+\n\
+.ResizeHandleActive {\n\
+  opacity: 1;\n\
 }\n\
\n\
```"}</tab>

<tab label="ColumnSelector.js">
{"```javascript\n\
import styles from './ColumnSelector.module.css';\n\
\n\
export default function ColumnSelector({ columns }) {\n\
  return (\n\
    <div className={styles.ColumnSelector}>\n\
      <div className={styles.Label}>Show Columns:</div>\n\
      <div className={styles.Checkboxes}>\n\
        {columns.map(column => (\n\
          <div key={column.id} className={styles.Checkbox}>\n\
            <label>\n\
              <input type='checkbox' {...column.getToggleHiddenProps()} />\n\
              {` ${column.Header}`}\n\
            </label>\n\
          </div>\n\
        ))}\n\
      </div>\n\
    </div>\n\
  );\n\
}\n\
\n\
```"}
</tab>

<tab label="ColumnSelector.module.css">
{"```css\n\
.ColumnSelector {\n\
  font-size: 1rem;\n\
  margin-bottom: 1rem;\n\
}\n\
\n\
.Label {\n\
  text-align: left;\n\
}\n\
\n\
.Checkbox {\n\
  text-align: left;\n\
}\n\
\n\
```"}
</tab>

</tabs>

Although column ordering also has a plugin, it doesn't provide props for easy controls like the others. As such, we'll skip it for this tutorial.

> Here's what it looks like so far:

Reviewer: ```suggestion
Here's what it looks like so far (click to interact):
```

Author: Done.

+  },\n\
+  function ScoreCell({ value }) {\n\
+    const color = value < 50 ? 'pink' : 'aquamarine';\n\
+    return <span style={{ color }}>{value}</span>;\n\
+  },\n\
+  function TemperamentCell({value}) {\n\
+    return <>{value}</>;\n\
+  },\n\
+];\n\
 \n\
 // react-table expects memoized columns and data, so we export a React hook to permit doing that.\n\
 export default function useData(numRows = 20, numCols = 5) {\n\
   const columns = React.useMemo(() => (\n\
     Array(numCols).fill(0).map((_, h) => {\n\
       const name = genericHeaders[h % genericHeaders.length];\n\
       const id = `col${h}`;\n\
       return {\n\
         Header: name,\n\
+        Cell: fancyCellRenderers[h % fancyCellRenderers.length],\n\
         accessor: id\n\
       };\n\
     })\n\
   ), [numCols]);\n\
 \n\
   const data = React.useMemo(() => (\n\
     Array(numRows).fill(0).map(() => {\n\
       const row = {};\n\
       for (let c = 0; c < numCols; c++) {\n\
         row[`col${c}`] = genericDataFuncs[c % genericDataFuncs.length]();\n\
       }\n\
       return row;\n\
     })\n\
   ), [numRows, numCols]);\n\
 \n\
   return { columns, data };\n\
 }\n\
\n\
```"}
</tab>
</tabs>

> With this feature, our demo is largely complete (interactive demo):

Reviewer: ```suggestion
With this feature, our demo is largely complete (click to interact):
```

Author: Done.
